janus_server/src/aggregator.rs

```diff
@@ -24,39 +192,135 @@ fn aggregator_filter(task_id: TaskId) -> BoxedFilter<(impl Reply,)> {
                 "max-age=86400",
             )
         })
-        .with(trace::named("hpke_config"))
-        .boxed()
+        .with(trace::named("hpke_config"));
+
+    let upload_endpoint = warp::path("upload")
+        .and(warp::post())
+        .and(with_shared_value(aggregator))
+        .and(with_decoded_message())
+        .and_then(
+            |aggregator: Arc<Mutex<Aggregator<C>>>, report: Report| async move {
+                let mut aggregator = aggregator.lock().await;
+
+                // Only the leader supports upload
+                if aggregator.role != Role::Leader {
+                    return Err(warp::reject::not_found());
+                }
+
+                aggregator
+                    .handle_upload(&report)
+                    .map_err(warp::reject::custom)?;
+
+                Ok(reply::with_status(warp::reply(), StatusCode::OK)) as Result<_, Rejection>
+            },
+        )
+        .with(trace::named("upload"));
+
+    Ok(hpke_config_endpoint.or(upload_endpoint).boxed())
 }
 
 /// Construct a PPM aggregator server, listening on the provided [`SocketAddr`].
 /// If the `SocketAddr`'s `port` is 0, an ephemeral port is used. Returns a
 /// `SocketAddr` representing the address and port the server are listening on
 /// and a future that can be `await`ed to begin serving requests.
-pub fn aggregator_server(
-    task_id: TaskId,
+pub fn aggregator_server<C: 'static + Clock>(
+    clock: C,
+    tolerable_clock_skew: Duration,
+    role: Role,
+    hpke_recipient: HpkeRecipient,
     listen_address: SocketAddr,
-) -> (SocketAddr, impl Future<Output = ()> + 'static) {
-    let routes = aggregator_filter(task_id).with(trace::request());
+) -> Result<(SocketAddr, impl Future<Output = ()> + 'static), Error> {
+    let routes = aggregator_filter(clock, tolerable_clock_skew, role, hpke_recipient)?
+        .with(trace::request());
```

Comment:
We are doubling up on warp tracing filters, with one here, and separate `trace::named` filters below. This results in the following output.

```
  2022-03-17T17:17:36.432990Z  INFO warp::filters::trace: finished processing with success, status: 200
    at /home/david/.cargo/registry/src/github.com-1ecc6299db9ec823/warp-0.3.2/src/filters/trace.rs:243 on ThreadId(2)
    in warp::filters::trace::context with hpke_config
    in warp::filters::trace::request with method: GET, path: /hpke_config, version: HTTP/1.1, remote.addr: 127.0.0.1:39378

  2022-03-17T17:17:36.433033Z  INFO warp::filters::trace: finished processing with success, status: 200
    at /home/david/.cargo/registry/src/github.com-1ecc6299db9ec823/warp-0.3.2/src/filters/trace.rs:243 on ThreadId(2)
    in warp::filters::trace::request with method: GET, path: /hpke_config, version: HTTP/1.1, remote.addr: 127.0.0.1:39378
```